hazelcast/src/main/java/com/hazelcast/internal/ascii/rest/HttpCommandProcessor.java

```diff
@@ -83,24 +102,167 @@
+     */
+    protected void prepareResponse(@Nonnull HttpCommand command,
+                                   @Nullable Object value) {
         if (value == null) {
             command.send204();
-        } else if (value instanceof byte[]) {
-            command.setResponse(CONTENT_TYPE_BINARY, (byte[]) value);
+        } else {
+            prepareResponse(RES_200, command, value);
+        }
+    }
+
+    /**
+     * Prepares a response with the provided status line and response value.
+     *
+     * @param statusLine the binary-encoded HTTP response status line
+     * @param command    the HTTP request
+     * @param value      the response value to send
+     */
+    protected void prepareResponse(@Nonnull byte[] statusLine,
+                                   @Nonnull HttpCommand command,
+                                   @Nonnull Object value) {
+        if (value instanceof byte[]) {
+            command.setResponse(statusLine, CONTENT_TYPE_BINARY, (byte[]) value);
         } else if (value instanceof RestValue) {
             RestValue restValue = (RestValue) value;
-            command.setResponse(restValue.getContentType(), restValue.getValue());
+            command.setResponse(statusLine, restValue.getContentType(), restValue.getValue());
         } else if (value instanceof HazelcastJsonValue || value instanceof JsonValue) {
-            command.setResponse(CONTENT_TYPE_JSON, stringToBytes(value.toString()));
+            command.setResponse(statusLine, CONTENT_TYPE_JSON, stringToBytes(value.toString()));
         } else if (value instanceof String) {
-            command.setResponse(CONTENT_TYPE_PLAIN_TEXT, stringToBytes((String) value));
+            command.setResponse(statusLine, CONTENT_TYPE_PLAIN_TEXT, stringToBytes((String) value));
         } else {
-            command.setResponse(CONTENT_TYPE_BINARY, textCommandService.toByteArray(value));
+            command.setResponse(statusLine, CONTENT_TYPE_BINARY, textCommandService.toByteArray(value));
+        }
+    }
+
+    /**
+     * Decodes HTTP post params contained in {@link HttpPostCommand#getData()}. The data
+     * should be encoded in UTF-8 and joined together with an ampersand (&).
+     *
+     * @param command            the HTTP post command
+     * @param expectedParamCount the number of parameters expected in the command
+     * @return the decoded params
+     * @throws UnsupportedEncodingException If character encoding needs to be consulted, but
+     *                                      named character encoding is not supported
+     */
+    private static String[] decodeParams(HttpPostCommand command, int expectedParamCount)
+            throws UnsupportedEncodingException {
+        final byte[] data = command.getData();
+        final String[] encoded = bytesToString(data).split("&");
+        final String[] decoded = new String[encoded.length];
+
+        if (encoded.length < expectedParamCount) {
+            return null;
```

Comment:
I wonder why you decided to handle error by returning `null`, not by throwing an exception. The disadvantage of this approach here is that if you forget to handle it anywhere, then you'll get `NullPointerException`.

Reply:
Right, I'll have to introduce some custom exception as not to get HTTP 500 from the post command handler. Would that be ok?

Reply:
Sounds good!

Reply:
Introduced exceptions for bad request and forbidden. Looks better now, thanks for the idea!